File: backend/refs/matlab/_aux_fun/likelihood_kalman_covid.m
function  loglik = likelihood_kalman_covid(y,x,theta,map_theta_to_mats,settings)

% Inputs

% y = T x n_y matrix of n_y measurment time series observations
% x = T x n_x matrix of n_x exogenous variables. This includes constants and lags.
% theta = n x 1 vector, includes all parameters.

%   State Equation
%   xi_{t+1} = F xi_{t} + v_{t+1}
%
%   Measurement Eq
%   y_t = A' x_t + H' xi_{t} + w_t
%
%   Error covariance
%   Q = E[v_t v_t']
%   R = E[w_t w_t']

%this function is application-specific, and maps the paramter vector theta to the state-space represtation
% given above.

fixed_param = settings.fixed_param;
covid_dummies = settings.covid_dummies;
matrices = map_theta_to_mats(theta,fixed_param,covid_dummies);

if settings.detrend_y == 1
    y(:,1) = y(:,1)-theta(5)*(1:length(y))';
    x(:,1) = x(:,1)-theta(5)*(0:length(y)-1)';
    x(:,2) = x(:,2)-theta(5)*(-1:length(y)-2)';
elseif settings.detrend_y == 2
    y(:,1) = y(:,1)-theta(5)*(1:length(y))';
    x(:,1) = x(:,1)-theta(5)*(0:length(y)-1)';
    x(:,2) = x(:,2)-theta(5)*(-1:length(y)-2)';
    x(:,3) = x(:,3)-theta(5)*(-2:length(y)-3)';
end


F = matrices.F;
A = matrices.A;
H = matrices.H;
Q = matrices.Q;
R = matrices.R;
kappa_seq = matrices.kappa_seq;

settings.smoother = 0;
% run the Kalman filter and obtain everything
output = kalman_filter_covid(y,x,F,A,H,Q,R,kappa_seq,settings);

% compute the log-likelihood.

loglik_vec = output.loglik_vec;
loglik = sum(loglik_vec); 
end

File: backend/pipeline/_aux_fun/likelihood_kalman_covid.py
from __future__ import annotations

from copy import deepcopy
from typing import Any

import numpy as np

from _aux_fun.kalman_filter_covid import kalman_filter_covid


def _get(obj: Any, name: str) -> Any:
    if isinstance(obj, dict):
        return obj[name]
    return getattr(obj, name)


def _set(obj: Any, name: str, value: Any) -> Any:
    if isinstance(obj, dict):
        obj[name] = value
    else:
        setattr(obj, name, value)
    return obj


def likelihood_kalman_covid(
    y: np.ndarray,
    x: np.ndarray,
    theta: np.ndarray,
    map_theta_to_mats,
    settings,
) -> float:
    """
    Compute the log-likelihood of the state-space model.

    This function:
    - maps theta to the state-space matrices
    - detrends the data if needed
    - runs the Kalman filter
    - returns the sum of loglik_vec
    """
    y_work = np.array(y, dtype=float, copy=True)
    x_work = np.array(x, dtype=float, copy=True)
    theta = np.asarray(theta, dtype=float)

    fixed_param = _get(settings, "fixed_param")
    covid_dummies = _get(settings, "covid_dummies")

    matrices = map_theta_to_mats(theta, fixed_param, covid_dummies)

    detrend_y = _get(settings, "detrend_y")
    T = y_work.shape[0]

    # MATLAB theta(5) -> Python theta[4]
    g = theta[4]

    if detrend_y == 1:
        y_work[:, 0] = y_work[:, 0] - g * np.arange(1, T + 1)
        x_work[:, 0] = x_work[:, 0] - g * np.arange(0, T)
        x_work[:, 1] = x_work[:, 1] - g * np.arange(-1, T - 1)

    elif detrend_y == 2:
        y_work[:, 0] = y_work[:, 0] - g * np.arange(1, T + 1)
        x_work[:, 0] = x_work[:, 0] - g * np.arange(0, T)
        x_work[:, 1] = x_work[:, 1] - g * np.arange(-1, T - 1)
        x_work[:, 2] = x_work[:, 2] - g * np.arange(-2, T - 2)

    F = _get(matrices, "F")
    A = _get(matrices, "A")
    H = _get(matrices, "H")
    Q = _get(matrices, "Q")
    R = _get(matrices, "R")
    kappa_seq = _get(matrices, "kappa_seq")

    # Run the Kalman filter with smoother turned off.
    settings_local = deepcopy(settings)
    _set(settings_local, "smoother", 0)

    output = kalman_filter_covid(
        y=y_work,
        x=x_work,
        F=F,
        A=A,
        H=H,
        Q=Q,
        R_base=R,
        kappa_seq=kappa_seq,
        settings=settings_local,
    )

    loglik_vec = _get(output, "loglik_vec")
    return float(np.sum(loglik_vec))